src/measure_theory/covering/differentiation.lean

```diff
@@ -0,0 +1,723 @@
+  refine le_antisymm _ bot_le,
+  calc μ s ≤ ∑' (x : h.index), μ (h.covering x) : h.measure_le_tsum
+  ... = ∑' (x : h.index), 0 : by { congr, ext1 x, exact h.covering_mem x.2 }
+  ... = 0 : by simp only [tsum_zero, add_zero]
+end
+
+/-- For every point `x`, sufficiently small sets in a Vitali family around `x` have finite measure.
+(This is a trivial result, following from the fact that the measure is locally finite). -/
+theorem eventually_measure_lt_top [is_locally_finite_measure μ] (x : α) :
+  ∀ᶠ a in v.filter_at x, μ a < ∞ :=
+begin
+  obtain ⟨ε, εpos, με⟩ : ∃ (ε : ℝ) (hi : 0 < ε), μ (closed_ball x ε) < ∞ :=
+    (μ.finite_at_nhds x).exists_mem_basis nhds_basis_closed_ball,
+  exact v.eventually_filter_at_iff.2 ⟨ε, εpos, λ a ha haε, (measure_mono haε).trans_lt με⟩,
+end
+
+/-- If two measures `ρ` and `ν` have, at every point of a set `s`, arbitrarily small sets in a
+Vitali family satisfying `ρ a ≤ ν a`, then `ρ s ≤ ν s` if `ρ ≪ μ`.-/
+theorem measure_le_of_frequently_le [sigma_compact_space α] [borel_space α]
+  {ρ : measure α} (ν : measure α) [is_locally_finite_measure ν]
+  (hρ : ρ ≪ μ) (s : set α) (hs : ∀ x ∈ s, ∃ᶠ a in v.filter_at x, ρ a ≤ ν a) :
+  ρ s ≤ ν s :=
+begin
+  -- this follows from a covering argument using the sets satisfying `ρ a ≤ ν a`.
+  apply ennreal.le_of_forall_pos_le_add (λ ε εpos hc, _),
+  obtain ⟨U, sU, U_open, νU⟩ : ∃ (U : set α) (H : s ⊆ U), is_open U ∧ ν U ≤ ν s + ε :=
+    exists_is_open_le_add s ν (ennreal.coe_pos.2 εpos).ne',
+  let f : α → set (set α) := λ x, {a | ρ a ≤ ν a ∧ a ⊆ U},
+  have h : v.fine_subfamily_on f s,
+  { apply v.fine_subfamily_on_of_frequently f s (λ x hx, _),
+    have := (hs x hx).and_eventually ((v.eventually_filter_at_mem_sets x).and
+      (v.eventually_filter_at_subset_of_nhds (U_open.mem_nhds (sU hx)))),
+    apply frequently.mono this,
+    rintros a ⟨ρa, av, aU⟩,
+    exact ⟨ρa, aU⟩ },
+  haveI : encodable h.index := h.index_countable.to_encodable,
+  calc ρ s ≤ ∑' (x : h.index), ρ (h.covering x) : h.measure_le_tsum_of_absolutely_continuous hρ
+  ... ≤ ∑' (x : h.index), ν (h.covering x) : ennreal.tsum_le_tsum (λ x, (h.covering_mem x.2).1)
+  ... = ν (⋃ (x : h.index), h.covering x) :
+    by rw [measure_Union h.covering_disjoint_subtype (λ i, h.measurable_set_u i.2)]
+  ... ≤ ν U : measure_mono (Union_subset (λ i, (h.covering_mem i.2).2))
+  ... ≤ ν s + ε : νU
+end
+
+section
+
+variables [sigma_compact_space α] [borel_space α] [is_locally_finite_measure μ]
+  {ρ : measure α} [is_locally_finite_measure ρ]
+
+/-- If a measure `ρ` is singular with respect to `μ`, then for `μ` almost every `x`, the ratio
+`ρ a / μ a` tends to zero when `a` shrinks to `x` along the Vitali family. This makes sense
+as `μ a` is eventually positive by `ae_eventually_measure_pos`. -/
+lemma ae_eventually_measure_zero_of_singular (hρ : ρ ⊥ₘ μ) :
+  ∀ᵐ x ∂μ, tendsto (λ a, ρ a / μ a) (v.filter_at x) (𝓝 0) :=
+begin
+  have A : ∀ ε > (0 : ℝ≥0), ∀ᵐ x ∂μ, ∀ᶠ a in v.filter_at x, ρ a < ε * μ a,
```

Comment:
Should this `have` be a lemma?

Reply:
Since ti's not gonna be used again, I don't think so. It would make sense to split it if the proof were really long or slow, but I don't think it's the case here.